Pull in Tags 3.0; Create Pages 3.1 FAT

volosied opened the pull request:
fixes #21427 

Updated the JSP/Pages Implementation to handle new Tag Libraries. The API already exists under io.openliberty.jakarta/tags.3.0.bnd. This PR 
- created the implementation bundle as well as other internal builds. 
- Renamed the new bundles to use tags over jstl.
- Updated com.ibm.ws.jsp implementation to process the new jakarta.tag.* URIs and transformed other jsp bundles using ee10 rules. 
- Enabled EE10 Repeat over the com.ibm.ws.jsp.2.3_fat bucket. - Skipped several EL tests due to changes in the Pages 3.1 API.  Issue to re-enabled them  ->  https://github.com/OpenLiberty/open-liberty/issues/21743
- Created the new Pages 3.1 FAT Bucket  (io.openliberty.pages.3.1.internal_fat)

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

pnicolucci commented on dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSP23JSP22ServerTest.java:
```diff
@@ -124,7 +124,7 @@ public void testJsp23to22FeatureChange() throws Exception {
      */
     @Test
-    @SkipForRepeat(SkipForRepeat.NO_MODIFICATION)
+    @SkipForRepeat({SkipForRepeat.NO_MODIFICATION, SkipForRepeat.EE10_FEATURES})
```
@volosied can you open an issue so we add an EE10 pages-3.1 test for this functionality? If you can provide a link here that would be good so I can ensure we track it. Thanks!

volosied replied:
This test should already be added under the new 3.1 FAT.

volosied replied:
Oops. It's not in this PR because I never pulled it in from my other branch. I'll add it in soon.

volosied replied:
Test added

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

tevans78 commented:
@volosied please remember to squash/fixup your commits before final delivery of this PR

tevans78 commented on dev/com.ibm.ws.jsp.2.3/transformed10.bnd:
```diff
@@ -8,11 +8,11 @@
-Bundle-SymbolicName: com.ibm.ws.jsp.2.3.jakarta
+Bundle-Name: WAS JSP 3.1 Jakarta
+Bundle-SymbolicName: com.ibm.ws.jsp.3.1.jakarta
 Bundle-Description: WAS JSP 2.3, version ${bVersion}; Jakarta Enabled
```
```suggestion
Bundle-Description: WAS JSP 3.1, version ${bVersion}; Jakarta Enabled
```

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied committed:
Implement Jakarta Tags 3.0 in Pages 3.1; Update JSP/Pages FATs
Create io.openliberty.org.apache.taglibs.standard.3.0

Create TLD only bundle: io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal

Update com.ibm.ws.jsp.2.3 to transform for EE10; Create 3.1 Pages Property file for internal versioning

Update com.ibm.ws.jsp implementation to use tags 3.0

Various updates: feature files; exact resources, bnd buildpaths, cnf depdendencies, and transformation rules

Update com.ibm.ws.jsp.2.3_fat for EE10

Create Pages 3.1 FAT: io.openliberty.pages.3.1.internal_fat

Refactoring to use tags over jstl in names; Review cleanup

Code Review Comments

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented on dev/io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal/.project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.jakarta.servlet.6.0</name>
```
Correct this

volosied replied:
Rename project to use tags 3.0

pnicolucci replied:
This looks good now!

volosied commented on dev/io.openliberty.org.apache.taglibs.standard.3.0/.project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.jakarta.servlet.6.0</name>
```
Correct this too

pnicolucci replied:
Looks good!

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

pnicolucci commented on dev/io.openliberty.jstl.2.0.facade/bnd.bnd:
```diff
@@ -0,0 +1,20 @@
+#*******************************************************************************
+<<<<<<< HEAD
+# Copyright (c) 2021, 2022 IBM Corporation and others.
```
This is a bad merge.

pnicolucci commented on dev/io.openliberty.jstl.2.0.facade/ee10.bnd:
```diff
@@ -0,0 +1,50 @@
+  org.apache.taglibs.standard.tag.rt.xml;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tei;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tlv;version="[3.0,4.0)", \
+  org.apache.taglibs.standard;version="[3.0,4.0)"
```
new line

pnicolucci commented on dev/io.openliberty.org.apache.taglibs.standard.3.0/bnd.bnd:
```diff
@@ -0,0 +1,41 @@
+#*******************************************************************************
+# Copyright (c) 2021 IBM Corporation and others.
```
New bnd should be a 2022 copyright date.

pnicolucci commented on dev/io.openliberty.org.apache.taglibs.standard.3.0/build.gradle:
```diff
@@ -0,0 +1,41 @@
+
+clean {
+    dependsOn deleteJars
+}
```
new line

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:
```diff
@@ -0,0 +1,189 @@
+@RunWith(FATRunner.class)
+public class TagsTestsEE10 {
+    private static final String APP_NAME = "TestTagsEE10";
+    // private static final String IMPORT_APP_NAME = "TestJSTLImport";
```
delete?

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * Tests Import of SQL Tag Library
+     * 
+     * TODO: Add DB to test tags? 
```
Do we have an issue opened?

volosied replied:
Created: https://github.com/OpenLiberty/open-liberty/issues/21719

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/publish/servers/tagsServer/server.xml:
```diff
@@ -0,0 +1,23 @@
+
+    <javaPermission className="java.util.PropertyPermission" name="dtm.debug" actions="read"/>
+
+     <javaPermission className="java.util.PropertyPermission" name="user.dir" actions="read"/>
```
Fix indentation please.

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/test-applications/TestTagsEE10.war/resources/WEB-INF/web.xml:
```diff
@@ -0,0 +1,20 @@
+    version="3.1"
+    xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd">
```
Why are we a 3.1 version webapp and not 6.0?

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.Locale;
+
+/**
+ *  Smoke Tests for JSTL EE10 Tags
```
Are these just the EE9 tests in the EE10 bucket?

https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSTLTests.java

If we repeat the JSP 2.3 bucket for EE10 why are these necessary? Are they different in some way?

volosied replied:
This EE10 class test the new jakarta.tags.* URIs.  The EE9 one was for the older sun URIs. 

I think the older URIs are also tested again EE10 in the com.ibm.ws.jsp2.3_fat (JSTLTests.java) because we kept backwards compatibility.

volosied replied:
But otherwise, the tests are exactly the same.

pnicolucci replied:
Can you add a comment with that information so when someone looks at this test class it is obvious to them?

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented on dev/io.openliberty.jstl.2.0.facade/ee10.bnd:
```diff
@@ -0,0 +1,50 @@
+  org.apache.taglibs.standard.functions;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.lang.jstl;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.lang.jstl.parser;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.lang.jstl.test;version="[3.0,4.0)";resolution:=optional, \
+  org.apache.taglibs.standard.lang.jstl.test.beans;version="[3.0,4.0)";resolution:=optional, \
```
These should two imports should actually be removed..

volosied commented:
Abandoned last build and restarted a new one: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_-CUUAAF-Ee2_YeU2rrO1jw

pnicolucci commented on dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java:
```diff
@@ -260,6 +261,7 @@ public void testEL30PropertyNotWritableException() throws Exception {
      */
     @Test
     @ExpectedFFDC("javax.el.PropertyNotWritableException")
+    @SkipForRepeat(SkipForRepeat.EE10_FEATURES)
```
In the PR comments you say these are skipped due to changes in the Pages API , can you point me to the issue where we'll investigate these and work to reenable / update the tests for the Pages 3.1 feature?

volosied replied:
https://github.com/OpenLiberty/open-liberty/issues/21743

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/FATSuite.java:
```diff
@@ -0,0 +1,45 @@
+import componenttest.rules.repeater.RepeatTests;
+
+/**
+ * Pages 3.0 Tests
```
3.1

pnicolucci commented on dev/io.openliberty.pages.3.1.internal_fat/test-applications/TestTagsEE10.war/resources/META-INF/permissions.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
copyright?

pnicolucci commented on dev/wlp-jakartaee-transform/rules/jakarta-renames.properties:
```diff
@@ -67,7 +67,6 @@ com.sun.xml.internal.messaging.saaj.util.transform=com.sun.xml.messaging.saaj.ut
 io.grpc=io.grpc
 io.grpc.netty=io.grpc.netty
 io.openliberty.cdi.spi=io.openliberty.cdi.spi
-jakarta.el=jakarta.el
```
Can you just explain to me why this is not necessary any longer?

volosied replied:
In the EL 5.0 PR, I needed to transform the "io.openliberty.jakarta.jstl.2.0" bundle in order to get features which depended on EL to resolve. 

By transforming o.openliberty.jakarta.jstl.2.0 (an EE9 bundle),  I needed the jakarta.el imports to update. However, the current transformation rules only applied to javax.el -> jakarta.el.  By stating jakarta.el=jakarta.el, the versions then updated. 

We have a new jstl 3.0 api, so this is rule is no longer necessary.

pnicolucci commented on dev/com.ibm.websphere.javaee.jsp.tld.2.2/bnd.bnd:
```diff
@@ -26,7 +26,7 @@ Export-Package: \
     org.apache.taglibs.standard.tag.rt.xml;version="1.0.0"
 
--includeresource: @${repo;com.ibm.ws.javax.j2ee:jsp.tld.standard;1.2}
+-includeresource: @${repo;com.ibm.ws.javax.j2ee:jsp.tld.standard;1.2;EXACT}
```
Can you explain the addition of EXACT?

volosied replied:
bnd treats the the version,  1.2  in this case, more as a minimum.  So if we uploaded version 2.0 for the same group id and archifact id, then it would pull this later version in instead. 

I ran into this problem with the jakarta tags 2.0 bundle after I uploaded the 3.0 jar.  

I just added EXACT to all the bnds for consistency.

pnicolucci commented on dev/com.ibm.ws.jsp.2.3_fat/bnd.bnd:
```diff
@@ -27,7 +27,12 @@ tested.features:\
     expressionLanguage-4.0,\
```
Copyright update?

pnicolucci commented on dev/io.openliberty.jstl.2.0.facade/original.bnd:
```diff
@@ -49,4 +46,4 @@ Import-Package: \
   org.apache.taglibs.standard.tei;version="[2.0,3.0)", \
   org.apache.taglibs.standard.tlv;version="[2.0,3.0)", \
-  org.apache.taglibs.standard;version="[2.0,3.0)"
+  org.apache.taglibs.standard;version="[2.0,3.0)"
```
newline

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

pnicolucci commented:
#libby

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
Squashed all the commits into one. Merging now..

volosied commented:
#libby
